index.bs

```diff
@@ -74,6 +75,125 @@ A user agent must not <a spec=HTML>delay the load event</a> of the <{link}> elem
+
+  1. Let |parsed| be the result of [=Parse JSON into Infra values|parsing JSON into Infra values=] given |input|.
+  1. If |parsed| is not a [=map=], then return null.
+  1. Let |result| be an empty [=speculation rule set=].
+  1. If |parsed|["`prefetch`"] [=map/exists=] and is a [=list=], then [=list/for each=] |prefetchRule| of |parsed|["`prefetch`"]:
+    1. If |prefetchRule| is not a [=map=], then [=iteration/continue=].
+    1. Let |rule| be the result of [=parse a speculation rule|parsing a speculation rule=] given |prefetchRule| and |baseURL|.
+    1. If |rule| is null, then [=iteration/continue=].
+    1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prefetch rules=].
+  1. Return |result|.
+</div>
+
+<div algorithm="parse a speculation rule">
+  To <dfn export>parse a speculation rule</dfn> given a [=map=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
+
+  1. If |input| has any [=map/key=] other than "`source`", "`urls`", and "`requires`", then return null.
+  1. If |input|["`source`"] does not [=map/exist=] or is not the [=string=] "`list`", then return null.
+  1. Let |urls| be an empty [=list=].
+  1. If |input|["`urls`"] does not [=map/exist=], is not a [=list=], or has any element which is not a [=string=], then return null.
+  1. [=list/For each=] |urlString| of |input|["`urls`"]:
+    1. Let |parsedURL| be the result of [=basic URL parser|parsing=] |urlString| with |baseURL|.
+    1. If |parsedURL| is failure, then [=iteration/continue=].
+    1. If |parsedURL|'s [=url/scheme=] is not an [=HTTP(S) scheme=], then [=iteration/continue=].
+    1. [=list/Append=] |parsedURL| to |urls|.
+  1. Let |requirements| be an empty [=ordered set=].
+  1. If |input|["`requires`"] [=map/exists=], but is not a [=list=], then return null.
+  1. [=list/For each=] |requirement| of |input|["`requires`"]:
+    1. If |requirement| is not the [=string=] "`anonymous-client-ip`", then return null.
+    1. [=set/Append=] |requirement| to |requirements|.
+  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls| and [=speculation rule/requirements=] |requirements|.
+</div>
+
+<h3 id="speculation-rules-processing">Processing Model</h3>
+
+Periodically, the user agent may [=queue a task=] to the [=DOM manipulation task source=] to [=consider speculation=] for any [=document=].
+
+<div class="note">
+  The user agent will likely do this after the insertion of new speculation rules, or when resources are idle and available.
+</div>
+
+<div algorithm="consider speculation">
+  To <dfn export>consider speculation</dfn> for a [=document=] |document|:
+
+  1. For each <{script}> element [=connected=] to |document| where <a spec=html>the script's type</a> is "`speculationrules`" and [=the script's result=] is a [=speculation rule set=]:
```

Comment:
My high-level feedback for this algorithm is a continuation of my comment above. I think we should model the way other scripts handle their "readiness", by marking the script as ready and adding the "result" to some per-document list of "available speculation rules" or something. That way we can just access that list of results here and use them instead of having to iterate through all connected scripts whose type = [...] and is result is non-null etc etc. Does that make sense? If that's unclear at all I'm happy to add what I'm thinking directly to this PR or help clarify what I'm getting at.

Also has per my one comment above, I think this will be necessary to avoid the "Otherwise" block from immediately executing the speculation rules scripts, assuming I'm reading that right.

Reply:
Sure, I was on the fence about whether to have an explicit list or not. I think I would like the ability to remove them which will require hooking the removal of script elements, but that's not hard to do. (And I don't really have to do it in this PR.)

Reply:
Cool, it seems like we have the "removal" part tracked by another TODO. I'll go ahead and resolve this. I will say it could be interesting to consider what happens if we remove a speculationrules script from one Document and move it to another. Normal scripts that do that don't execute (they have to execute on the document they were prepared on), but I'm not sure if we want to consider doing the same thing here or not. Import maps doesn't seem to care.